### PR TITLE
br/restore: add version check for backup schema_version

### PR DESCRIPTION
**Pull request by YuJuncen**
### What problem does this PR solve? 
a special fix for v4.0 versions of #790 

Because #685 cannot be cherry-picked to release-4.0, 4.0.x BR would corrupt 4.0 clusters without even a warn message when the backup archive come from a table with non-int clustered primary key. This PR would add a check to the backup archive version, and intercept restoring from incompatible backup archives.

### What is changed and how it works?
CheckClusterVersion was changed to check version in different way.
And added a check to cluster version stored in backupmeta during restore. (after reading backup meta)

**Currently, this check only checks major versions (because of clustered index).**

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)

Code changes

 - Has exported function/method change
 signature change for `CheckClusterVersion`.
``` diff
- CheckClusterVersion(ctx context.Context, client pd.Client) error
+ CheckClusterVersion(ctx context.Context, client pd.Client, checker VerChecker) error
```

### Release Note

 - BR would check cluster version of backup now.

### CODE REVIEW

File: pkg/version/version.go [hunk 1]
```diff
@@ -6,15 +6,17 @@ import (
 	"context"
 	"fmt"
 	"regexp"
+	"strconv"
 	"strings"
 
+	"github.com/pingcap/log"
+	pd "github.com/tikv/pd/client"
+	"go.uber.org/zap"
+
```
Comment (overvenus) [inline pkg/version/version.go:15]:
```suggestion
```

File: pkg/version/version.go [hunk 2]
```diff
@@ -6,15 +6,17 @@ import (
...
 	"github.com/pingcap/kvproto/pkg/metapb"
```
Comment (overvenus) [inline pkg/version/version.go:17]:
```suggestion
```

File: pkg/version/version.go [hunk 3]
```diff
@@ -199,3 +224,20 @@ func CheckTiDBVersion(versionStr string, requiredMinVersion, requiredMaxVersion
 	}
 	return CheckVersion("TiDB", *version, requiredMinVersion, requiredMaxVersion)
 }
+
+// NormalizeBackupVersion normalizes the version string from backupmeta.
+func NormalizeBackupVersion(version string) *semver.Version {
```
Comment (overvenus) [inline pkg/version/version.go:229]: Seem `version.removeVAndHash` can be replaced by this funcation.

**Review by overvenus — COMMENTED** [review]
rest LGTM

File: pkg/version/version.go (hunk 3 above)
Comment (YuJuncen) [reply pkg/version/version.go:229]: You mean merge those two functions? Seems `removeVAndHash` focus on the version string made by the make script, and `NormalizeBackupVersion` focus on the version string returned from TiKV. Maybe it's hard to just replace the former with the latter. 🤔

**Comment by YuJuncen** [comment]
/run-integration-tests

port in use

File: pkg/version/version.go [hunk 4]
```diff
@@ -99,44 +99,67 @@ func CheckClusterVersion(ctx context.Context, client pd.Client) error {
...
+	return nil
+}
 
-		if tikvVersion.Major != BRVersion.Major {
-			return errors.Annotatef(berrors.ErrVersionMismatch, "TiKV node %s version %s and BR %s major version mismatch, please use the same version of BR",
-				s.Address, tikvVersionString, build.ReleaseVersion)
+// CheckVersionForBackup checks the version for backup and
+func CheckVersionForBackup(backupVersion *semver.Version) VerChecker {
+	return func(store *metapb.Store, ver *semver.Version) error {
+		if backupVersion.Major != ver.Major {
```
Comment (3pointer) [inline pkg/version/version.go:116]: How about change it to `backupVersion.Major < ver.Major`?

Comment (YuJuncen) [reply pkg/version/version.go:116]: sounds good

**Comment by YuJuncen** [comment]
listen tcp 0.0.0.0:10080: bind: address already in use

/run-integration-tests

**Review by 3pointer — COMMENTED** [review]
/lgtm

**Comment by YuJuncen** [comment]
[2021-05-08T11:29:42.533Z] [2021/05/08 19:29:40.505 +08:00] [INFO] [restore_raw.go:110] ["all files are filtered out from the backup archive, nothing to restore"]

/run-integration-tests

**Pull request merged**